packages/hash-bundle/README.md

```diff
@@ -0,0 +1,12 @@
+# hash-bundle
+
+`hashBundle` is a function that takes a bundle and returns a consistent hash of the bundle that can be used to perform an integrity check of the bundle.
+To that end, `hashBundle` verifies the internal consistency, completeness, coherence, and conciseness (no extra files) of the bundle.
+
+```js
+const bundle = await bundleSource('path/to/bundle.js');
+// 'bundle' is JSON-serializable
+const version = await hashBundle(bundle);
```

Comment:
Let's call it a `bundleID`, since `version` implies (to me) a scalar integer of a class, not an instance.

In my prototype, I used `b1-` as a prefix (for "bundle"). That speaks to a larger application in which this is embedded, where we're anticipating the need to distinguish between modules (`m1-`?) and full bundles, and maybe some other user-visible identifiers that we haven't named yet. I kind of want to ask `hash-bundle` to use `b1-`, but that might be imposing an application-specific requirement on an upstream library, which feels wrong.

If `hash-bundle` uses `v1-`, I might have swingset prepend a `b` (so users see `bv1-HASH`), or replace the `v1-` with `b1-`.. not sure.

This document seems as good a place as any to specify what the string really is, since anyone using this package will care about the details (in particular they'll care about the details not changing, since this is a compatibility guarantee). Maybe some text like:

* "getExport" -type bundles: the BundleID is a `v0-` prefix followed by the lowercase hex SHA512 hash of the full contents. Note that this includes platform-native line endings, so the same source code bundle into different IDs.
* "endoZipBase64"-type bundles (the current default): the BundleID is a `v1-` prefix followed by the lowercase hex SHA512 hash of the compartment map. This does not depend upon the bundling platform.

Reply:
Oh, and the README should make it clear that calling `hashBundle(bundle)` does not verify that the bundle is internally consistent, but merely computes an ID that might match the rest of the bundle. Maybe provide an example of what needs to be done to compute+validate the ID of an alleged bundle (which requires an additional package).

For swingset and chain deploy, the use cases are:
* contract/vat/bundle author writes their code, then pushes the "upload/deploy" button
  * their agent bundles the sources, computes the ID, performs an upload
  * author needs to be told the ID
  * to avoid surprises / guard against bugs, the agent should probably validate the bundle against the computed ID, but it's not strictly necessary
    * worst case (platform-sensitive line-ending bug??) a bug in the bundling/hashing process would cost the author fees when they attempt to upload and get rejected for a mismatch
* the chain receives an alleged bundle and ID, and wants to store it
  * chain must validate the complete bundle against the claimed ID
  * some environments might not require the client to supply an ID, so the chain may need to compute the ID first
* later, when loading the bundle, it's probably a good idea to re-validate it, just before/during the `eval`

Reply:
Hm, now I'm on the fence at to whether I want `hash-bundle` to also validate the whole thing, and/or if there is a use case for only hashing the compartment map. In general I want `hash-bundle` to actually look at the whole bundle, but making it do everything would preclude it being used in a place where, I dunno, maybe you want to quickly/cheaply tell that a given bundle is not a duplicate of something you already have, or you want to look up previously-calculated information about an ID.

Oh, ok, so when we get to uploading bundles efficiently (just the modules they don't already have), the uploading agent (running `agoric install-contract BUNDLE` or something) might have a previously-created bundle and is willing to rely upon its correctness. It wants to quickly/cheaply learn the ID so it can send the first message to the chain ("I'd like to upload bundle-ID, how much of it do you already have?"), and they don't need to validate anything else about their local bundle to do that.

That use case will also want a function that takes a reliable bundle and returns a set of module IDs, probably at the same time as returning the overall bundle (compartment map) ID. We might compute and stash that during bundling, but we'll probably need to apply it to a pre-made bundle, and then we're either willing to rely upon the integrity or not. If we are, then all we have to do is hash the compartment map and return both the bundle hash and the hashes from inside the compartment map. If not, we need to validate everything, and we should gather the hashes to return at the same time, instead of hashing twice.

Reply:
We converged on a solution that avoids naming a “tagged hash” at this layer of our stack. A bundle will contain an `endoZipBase64Sha512` property if it has an associated hash. `bundleSource` will produce this property. `checkBundle` will verify its consistency with the content of `endoZipBase64`.